### PR TITLE
Send Amount + Address input fields

### LINKED ISSUE #245: Add Validation Regex that support localization
`secant/Util/Strings.swift` file has placeholder regex's for reference. They should be replace for ones that support localization. Specifically on number validation.

### LINKED ISSUE #294: [Scaffold] Send Screen - transaction + address inputs
Adam created the component for the generic input with the title and accessory views. Integrate this component into the send screen for both 
- amount (zec/usd)
- address

with the following features:

**amount (zec/usd)**
1. [x] max value (done but hardcoded)
2. [x] clear value (missing completely)
3. [x] switch zec/usd (review)
4. [x] validatate amount (review)
5. [x] currency symbol preceding the value ($ <value> or ZEC symbol <value>), right now the $ is integrated to the string, support for ZEC symbol is missing completely

**address** - this needs to be done completely, generics of the field allows us to make it pretty easily
1. [x] clear
2. [x] truncate in the middle
3. [x] validate address

I expect to iterate over the Adam's code a little bit, at least:
1. [x] ready to be composable (it's not as not everything is Equatable)
2. [x] localization support
3. [x] features support (there's no max to clear switching at the moment, etc.)
4. [x] replace all hardcoded values with injected ones
5. [x] add missing unit tests

### PR DESCRIPTION
**Pull request by LukasKorba**
Closes #294 
Closes #245

This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
- [x] Does the code abide by the [Coding Guidelines](../blob/main/docs/CODING_GUIDELINES.md)?
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [x] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
- [ ] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), etc.)
- [x] Run the app: Did you run the app and try the changes? 
- [ ] Did you provide Screenshots of what the App looks like before and after your changes as part of the description of this PR? (only applicable to UI Changes)
- [x] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.

# Reviewer

- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
- [ ] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?_You will find manual testing guidelines under our [manual testing section](../blob/main/docs/testing/manual_testing)_
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage before and after your changes and when possible, leave it in a better place. [Learn More...](../blob/main/docs/testing/local_coverage.md)_
- [ ] Documentation: Did you review Docs, [README.md](../blob/main/README.md), [LICENSE.md](../blob/main/LICENSE.md), and [Architecture.md](../blob/main/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the app and try the changes? While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data.

### CODE REVIEW

File: secant/Features/Send/SendStore.swift [hunk 1]
```diff
@@ -37,7 +37,13 @@ struct SendState: Equatable {
     var isSendingTransaction = false
     var totalBalance = 0.0
     var transaction: Transaction
-    var transactionInputState: TransactionInputState
+    var transactionAddressInputState: TransactionAddressInputState
+    var transactionAmountInputState: TransactionAmountInputState
+    
+    var isValidForm: Bool {
```
Comment (pacu) [inline secant/Features/Send/SendStore.swift:43]: this can seem a silly test to cover but it could lead to avoidable user errors further in the flow

**Review by pacu — CHANGES_REQUESTED** [review]

File: secant/Features/Send/Views/CreateTransactionView.swift [hunk 2]
```diff
@@ -13,40 +13,46 @@ struct CreateTransaction: View {
...
+                            state: \.transactionAddressInputState,
+                            action: SendAction.transactionAddressInput
+                        )
                     )
-                        .font(.system(size: 14))
-                        .padding()
-                        .background(Color.white)
-                        .foregroundColor(Asset.Colors.Text.importSeedEditor.color)
+                    
+                    if !viewStore.transactionAddressInputState.isValidAddress &&
```
[the remark this answers is not in the document]
Comment (LukasKorba) [reply secant/Features/Send/Views/CreateTransactionView.swift:47]: added computed variable

File: secant/Features/Send/Views/CreateTransactionView.swift [hunk 3]
```diff
@@ -13,40 +13,46 @@ struct CreateTransaction: View {
...
-                                .textFieldTitleAccessoryButtonStyle
-                        },
-                        inputAccessoryView: {
-                        }
+                            state: \.transactionAmountInputState,
+                            action: SendAction.transactionAmountInput
+                        )
                     )
+                    
+                    if !viewStore.transactionAmountInputState.textFieldState.valid &&
```
[the remark this answers is not in the document]
Comment (LukasKorba) [reply secant/Features/Send/Views/CreateTransactionView.swift:27]: added computed variable

File: secant/Features/Send/Views/TransactionConfirmationView.swift [hunk 4]
```diff
@@ -6,7 +6,7 @@ struct TransactionConfirmation: View {
 
     var body: some View {
         VStack {
-            Text("Send \(String(format: "%.7f", Int64(viewStore.transactionInputState.amount).asHumanReadableZecBalance())) ZEC")
+            Text("Send \(String(format: "%.7f", Int64(viewStore.transaction.amount).asHumanReadableZecBalance())) ZEC")
```
[the remark this answers is not in the document]
Comment (LukasKorba) [reply secant/Features/Send/Views/TransactionConfirmationView.swift:9]: refactored at all places

**Review by pacu — COMMENTED** [review]
LGTM!

I understand that the Tests are WIP. so approval is pending of having them

File: secant/UIComponents/TextFields/TransactionAddress/TransactionAddressInputStore.swift [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    
+    private static let addressReducer = TransactionAddressInputReducer { state, action, environment in
+        switch action {
+        case .clearAddress:
+            state.textFieldState.text = ""
+            return .none
+
+        case .textField(.set(let address)):
+            do {
+                if try environment.wrappedDerivationTool.isValidTransparentAddress(address) {
```
[the remark this answers is not in the document]
Comment (LukasKorba) [reply secant/UIComponents/TextFields/TransactionAddress/TransactionAddressInputStore.swift]: tests implemented

File: secant/Features/Send/SendStore.swift (hunk 1 above)
Comment (LukasKorba) [reply secant/Features/Send/SendStore.swift:48]: tests implemented

**Review by pacu — CHANGES_REQUESTED** [review]
Great Work Lukas. Please check the remarks about the Tests.

File: secant/UIComponents/TextFields/TransactionAmount/CurrencySelectionStore.swift [hunk 6]
```diff
@@ -35,7 +35,7 @@ enum CurrencySelectionAction: Equatable {
     case swapCurrencyType
```
[the remark this answers is not in the document]
Comment (LukasKorba) [reply secant/UIComponents/TextFields/TransactionAmount/CurrencySelectionStore.swift:35]: true, I created an issue to handle exactly this https://github.com/zcash/secant-ios-wallet/issues/315

File: secantTests/SendTests/TransactionAmountInputTests.swift [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+
+// TODO: these tests will be updated with the Zatoshi/Balance representative once done, issue #272 https://github.com/zcash/secant-ios-wallet/issues/272
+
+class TransactionAmountInputTests: XCTestCase {
+    func testMaxValue() throws {
+        try XCTSkipUnless(Locale.current.regionCode == "US", "testMaxValue is designed to test US locale only")
+
+        let store = TestStore(
+            initialState:
+                TransactionAmountInputState(
```
[the remark this answers is not in the document]
Comment (LukasKorba) [reply secantTests/SendTests/TransactionAmountInputTests.swift:20]: already ticketed, I'm not sure if we can rely on XCUITest when we do unit tests but a dependency should solve the issue for us (https://github.com/zcash/secant-ios-wallet/issues/312)

File: secantTests/SendTests/SendTests.swift [hunk 8]
```diff
@@ -117,4 +125,300 @@ class SendTests: XCTestCase {
...
+    }
+    
+    func testInvalidAddressFormatEmptyInput() throws {
+        let store = TestStore(
+            initialState: .placeholder,
+            reducer: SendReducer.default,
+            environment: sharedTestEnvironment
+        )
+
+
```
[the remark this answers is not in the document]
Comment (LukasKorba) [reply secantTests/SendTests/SendTests.swift:187]: fixed

File: secantTests/SendTests/SendTests.swift [hunk 9]
```diff
@@ -103,7 +111,7 @@ class SendTests: XCTestCase {
             state.isSendingTransaction = true
         }
 
-        testScheduler.advance(by: 0.01)
+        Self.testScheduler.advance(by: 0.01)
```
[the remark this answers is not in the document]
Comment (LukasKorba) [reply secantTests/SendTests/SendTests.swift:114]: I made it independent

File: secantTests/SendTests/SendTests.swift [hunk 10]
```diff
@@ -10,9 +10,21 @@ import XCTest
 import ComposableArchitecture
 import ZcashLightClientKit
 
+// TODO: these tests will be updated with the Zatoshi/Balance representative once done, issue #272 https://github.com/zcash/secant-ios-wallet/issues/272
+
+// swiftlint:disable type_body_length
 class SendTests: XCTestCase {
+    static let testScheduler = DispatchQueue.test
```
[the remark this answers is not in the document]
Comment (LukasKorba) [reply secantTests/SendTests/SendTests.swift:17]: I made it independent

**Review by pacu — APPROVED** [review]

**Pull request merged**